### PR TITLE
stepstoreproduce model

### PR DESCRIPTION
**Pull request by ayush-1506**
Fixes : #19.

I think we'd also need to manually add more examples here. There aren't get enough labels we get for this classifier.

`X: (2462, 62431), y: (2462,)`

### CODE REVIEW

**Comment by marco-c** [comment]
What are the accuracy, precision and recall?

File: bugbug/models/stepstoreproduce.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class StepsToReproduceModel(BugModel):
+    def __init__(self, lemmatization=False):
+        BugModel.__init__(self, lemmatization)
+
+        self.sampler = RandomUnderSampler(random_state=0)
+
+        feature_extractors = [
+            bug_features.has_str(),
```
Comment (marco-c) [inline bugbug/models/stepstoreproduce.py]: We shouldn't use this as a feature, because it's the label

File: bugbug/models/stepstoreproduce.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    ),
+                ),
+            ]
+        )
+
+        self.clf = xgboost.XGBClassifier(n_jobs=16)
+        self.clf.set_params(predictor="cpu_predictor")
+
+    def rollback(self, change):
+        return change["removed"].startswith("stepswanted")
```
Comment (marco-c) [inline bugbug/models/stepstoreproduce.py]: Why are you rolling back the bug to when stepswanted was removed?

**Comment by ayush-1506** [comment]
```
Cross Validation scores:
Accuracy: f0.6794171294569649 (+/- 0.1805352640113031)
Precision: f0.9883502752225368 (+/- 0.007868684083782445)
Recall: f0.6791752599467217 (+/- 0.19161382476548508)
```

File: bugbug/models/stepstoreproduce.py (hunk 1 above)
Comment (ayush-1506) [reply bugbug/models/stepstoreproduce.py]: Oh, right. Missed this.

**Comment by marco-c** [comment]
> Cross Validation scores:

We need the scores after the removal of has_str feature, otherwise the model is too easy.

**Comment by ayush-1506** [comment]
Yes, these are the scores after removing has_str.

**Comment by marco-c** [comment]
> Yes, these are the scores after removing has_str.

Not as bad as I was expecting. Could you show me the confusion matrix too? Actually, post the whole output with the metrics and matrices for all confidence intervals.

**Comment by ayush-1506** [comment]
Sure:
```
>>> python run.py --train --goal=stepstoreproduce

X: (2462, 62429), y: (2462,)
Cross Validation scores:
Accuracy: f0.6794171294569649 (+/- 0.1805352640113031)
Precision: f0.9883502752225368 (+/- 0.007868684083782445)
Recall: f0.6791752599467217 (+/- 0.19161382476548508)
X_train: (116, 62429), y_train: (116,)
X_test: (247, 62429), y_test: (247,)

Top 18 Features:
1. 'comments__you' (-0.08424823731184006)
2. 'comments__regression' (+0.08394227176904678)
3. 'comments__should' (+0.06826793402433395)
4. 'comments__please' (-0.06098382547497749)
5. 'comments__the' (+0.05945780128240585)
6. 'comments__don' (-0.05036022141575813)
7. 'comments__as' (-0.0483185313642025)
8. 'comments__if' (-0.04371195659041405)
9. 'comments____url__' (+0.038325030356645584)
10. 'comments__profile' (-0.02546454593539238)
11. 'comments__false' (-0.023365214467048645)
12. 'comments__str' (+0.02319621853530407)
13. 'comments__on' (+0.023011967539787292)
14. 'comments__in' (+0.020338984206318855)
15. 'comments__report' (-0.015745362266898155)
16. 'comments__attachment' (+0.014188256114721298)
17. 'comments__my' (-0.013193286024034023)
18. 'comments__expected' (-0.013154595158994198)
No confidence threshold - 247 classified
[[166  69]
 [  4   8]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.71      0.67      0.82      0.69      0.47       235
          0       0.10      0.67      0.71      0.18      0.69      0.47        12

avg / total       0.93      0.70      0.67      0.79      0.69      0.47       247


Confidence threshold > 0.6 - 212 classified
[[155  46]
 [  3   8]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.98      0.77      0.73      0.86      0.75      0.56       201
          0       0.15      0.73      0.77      0.25      0.75      0.56        11

avg / total       0.94      0.77      0.73      0.83      0.75      0.56       212


Confidence threshold > 0.7 - 187 classified
[[144  36]
 [  1   6]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.80      0.86      0.89      0.83      0.68       180
          0       0.14      0.86      0.80      0.24      0.83      0.69         7

avg / total       0.96      0.80      0.86      0.86      0.83      0.68       187


Confidence threshold > 0.8 - 146 classified
[[122  18]
 [  1   5]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.87      0.83      0.93      0.85      0.73       140
          0       0.22      0.83      0.87      0.34      0.85      0.72         6

avg / total       0.96      0.87      0.83      0.90      0.85      0.73       146


Confidence threshold > 0.9 - 87 classified
[[76  8]
 [ 1  2]]
                   pre       rec       spe        f1       geo       iba       sup

          1       0.99      0.90      0.67      0.94      0.78      0.62        84
          0       0.20      0.67      0.90      0.31      0.78      0.59         3

avg / total       0.96      0.90      0.67      0.92      0.78      0.62        87

```

**Comment by marco-c** [comment]
OK, thanks!
So the precision for 0 (absence of STR) is pretty bad. We definitely need to add more examples.

File: bugbug/models/stepstoreproduce.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                elif bug_data["cf_has_str"] == "yes":
+                    classes[int(bug_data["id"])] = 1
+            elif "stepswanted" in bug_data["keywords"]:
+                classes[int(bug_data["id"])] = 0
+            else:
+                for entry in bug_data["history"]:
+                    for change in entry["changes"]:
+                        if change["removed"].startswith("stepswanted"):
+                            classes[int(bug_data["id"])] = 1
+
```
Comment (marco-c) [inline bugbug/models/stepstoreproduce.py]: Could you log the number of 1 and 0? Similarly to what we do in bugbug/models/defect_enhancement_task.py

Comment (ayush-1506) [reply bugbug/models/stepstoreproduce.py]: Sure. We get 70 bugs with no steps and 2392 bugs with steps.

File: bugbug/models/stepstoreproduce.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        print(
+            "{} bugs have no steps to reproduce".format(
+                sum(1 for label in classes.values() if not label)
```
Comment (marco-c) [inline bugbug/models/stepstoreproduce.py]: Nit: `if label == 0`, better to be explicit

File: bugbug/models/stepstoreproduce.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            )
+        )
+        print(
+            "{} bugs have steps to reproduce".format(
+                sum(1 for label in classes.values() if label)
```
Comment (marco-c) [inline bugbug/models/stepstoreproduce.py]: Same here, `if label == 1`

File: bugbug/models/stepstoreproduce.py (hunk 3 above)
Comment (marco-c) [reply bugbug/models/stepstoreproduce.py]: What is the output? How many do we have?

Comment (marco-c) [reply bugbug/models/stepstoreproduce.py]: OK, thanks!

**Pull request merged**